kpatch-build/create-diff-object.c

```diff
@@ -771,6 +775,86 @@ static bool kpatch_line_macro_change_only(struct section *sec)
 
 	return true;
 }
+
+#elif __s390x__
+
+static int insn_get_length(unsigned char *inst)
+{
+	switch (inst[0] >> 6) {
+		case 0b00:
+			return 2;
+		case 0b01:
+		case 0b10:
+			return 4;
+		case 0b11:
+		default:
+			return 6;
+	}
+}
+
+static bool kpatch_line_macro_change_only(struct section *sec)
+{
+	unsigned char *start1, *start2;
+	unsigned long size, offset, length;
+	struct rela *rela;
+	int lineonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return false;
+
+	start1 = sec->twin->data->d_buf;
+	start2 = sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += length) {
+		length = insn_get_length(start1 + offset);
+
+		if (insn_get_length(start1 + offset) != insn_get_length(start2 + offset))
+			return false;
+
+		if (!memcmp(start1 + offset, start2 + offset, length))
+			continue;
+
+		/* Verify lghi %r4 <line number> */
+		if ((start1[offset] != 0xa7) || (start1[offset+1] != 0x49) ||
+		    (start2[offset] != 0xa7) || (start2[offset+1] != 0x49))
+			return false;
+
+		/*
+		 * Verify zero or more string relas followed by a
+		 * warn_slowpath_* or another similar rela.
+		 */
+		found = 0;
+		list_for_each_entry(rela, &sec->rela->relas, list) {
+			if (rela->offset < offset + length)
+				continue;
+			if (rela->string)
+				continue;
+			if (!strncmp(rela->sym->name, "__warned.", 9) ||
+			    !strncmp(rela->sym->name, "__already_done.", 15))
```

Comment:
Addition of .__func__ is necessary

Reply:
I have a `__func__` check coming soon in #1254.